### PR TITLE
[WIP] N*G Triton group gemm for MoE 

### PR DESCRIPTION
**Pull request by lessw2020**
This PR adds a Triton Group GEMM with full backwards pass support, for integration with MoE training. 
The forward pass is from FBGemm experimental:
https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/experimental/gemm/triton_gemm/grouped_gemm.py

A. - Forward unit tests expanded and refactored.  now includes DeepSeek v3 main shapes:
~~~
Testing BF16 N*G GroupGeMM with G=1, M=64
Testing BF16 N*G GroupGeMM with G=1, M=512
Testing BF16 N*G GroupGeMM with G=1, M=2048
Testing BF16 N*G GroupGeMM with G=4, M=64
Testing BF16 N*G GroupGeMM with G=4, M=512
Testing BF16 N*G GroupGeMM with G=4, M=2048
Testing BF16 N*G GroupGeMM with G=16, M=64
Testing BF16 N*G GroupGeMM with G=16, M=512
Testing BF16 N*G GroupGeMM with G=16, M=2048
.Testing BF16 N*G Deepseek shape: G=4, M=8192, N=4096, K=7168
Testing BF16 N*G Deepseek shape: G=4, M=8192, N=7168, K=2048
Testing BF16 N*G Deepseek shape: G=8, M=4096, N=4096, K=7168
Testing BF16 N*G Deepseek shape: G=8, M=4096, N=7168, K=2048
.
----------------------------------------------------------------------
Ran 2 tests in 75.435s

OK
~~~

1 - Backward Numerics with BF16 have been verified on sample sizes and the core DeepSeek v3 shapes. 
~~~
2025-03-13 16:40:08,294 - INFO - Gradient shapes - grad_x: torch.Size([1024, 256]), grad_w: torch.Size([10240, 256])
2025-03-13 16:40:08,294 - INFO - Running PyTorch reference implementation
2025-03-13 16:40:08,591 - INFO - Comparing gradients with PyTorch reference
2025-03-13 16:40:08,602 - INFO - Maximum gradient error - grad_x: 0.125, grad_w: 0.125
2025-03-13 16:40:08,641 - INFO - ✓ SUCCESS! grad_X matches the PyTorch reference (allclose check passed)
2025-03-13 16:40:08,641 - INFO - ✓ SUCCESS! grad_W matches the PyTorch reference (allclose check passed)
2025-03-13 16:40:08,641 - INFO - Gradients allclose check - grad_x: True, grad_w: True
2025-03-13 16:40:08,641 - INFO - ✓ SUCCESS: Gradients match the PyTorch reference (allclose check passed)
~~~

2 - **Todos:**
a - fp8 support
b - TMA (used in forward now but not in backward) 
c - WS
d - Perf and auto-tuning

3 - Integration - ready now for BF16 though may want to do perf work first.

### CODE REVIEW

File: torchtitan/experiments/kernels/triton_group_gemm/tgroup_gemm_forward.py [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+"""
+
+
+def grouped_gemm_forward(
+    x: torch.Tensor, w: torch.Tensor, m_sizes: torch.Tensor
+) -> torch.Tensor:
+    return _grouped_gemm(x, w, m_sizes)
+
+
+def grouped_gemm_fp8_rowwise(
```
Comment (GD06) [inline torchtitan/experiments/kernels/triton_group_gemm/tgroup_gemm_forward.py]: I am curious about why these wrapper functions cannot call FBGEMM's kernels directly, but you have to deepcopy Triton impls of the _grouped_gemm from FBGEMM?
 https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/experimental/gemm/triton_gemm/grouped_gemm.py

cc @jianyuh  @yuchenhao @levendlee

Comment (lessw2020) [reply torchtitan/experiments/kernels/triton_group_gemm/tgroup_gemm_forward.py]: Hi @GD06  - thanks for the question.  A number of reasons why we are using the FBGemm forward portion as standalone version:

1 - Putting a requirement on building FBGEMM for a user to use maybe use two forward pass kernels is a heavy dependency and for both Titan and Tune, we want to minimize external dependencies.   
I already discussed having an FBGemm dependency with Tune (based on talking with Josh who was in favor of us taking an FBGemm dependency) and they declined.  So that was generally a non-starter. 
 
Also, for reference, not sure if you have reviewed FBGemm documentation recently, but it's both very out of date and not exactly supportive of taking a dependency: 
"The FBGEMM_GPU CUDA package is currently only built for the **SM70 and SM80** architectures (V100 and A100 GPUs respectively). _**Support for other architectures can be achieved by building the package from scratch, but is not guaranteed to work**_..."

2 - The two forward passes here are from an experimental section, meaning there's no guarantee it would stick around.  Thus, a sudden change/removal could break users at any point vs if we deepcopy it in we know it will be available (at the tradeoff of being out of sync if you made future updates). 

 FBGEmm doesn't supply any backward pass support afaik (hence this PR), and for training we need that. This PR adds backward pass support to these, and I wanted to make sure we had a full training support where it was a standalone impl not subject to future sudden breakage. 

3 - minor point - I had already added some clarifying comments to the bf16 forward b/c N was confusing as to meaning N*G or N, removed some un-needed HIP related configs and code, etc. These are not changes you would want (well maybe the clarifying comments) but is helpful for us to keep focused and maintainable code.   I have no idea what the PR process/time needed would be like for FBGemm if we have updates to the forward pass and that's another reason to keep it standalone b/c we can make changes rapidly if needed.

4 - Lastly, it's likely that what we really need is an M*G group gemm rather than N*G for better supporting all2all MoE, so probably a moot point overall, as I've already started an M*G forward bf16 impl and will PR that shortly in which case there is no linkage regardless.  Unfortunately, this PR will likely be dropped and replaced with the M*G (though if you plan to make an M*G please let me know).  

Hope that helps clarify the reasoning here and happy to discuss internally as well.

Comment (GD06) [reply torchtitan/experiments/kernels/triton_group_gemm/tgroup_gemm_forward.py]: Thanks for the reply! 

> FBGEmm doesn't supply any backward pass support afaik (hence this PR), and for training we need that. This PR adds backward pass support to these, and I wanted to make sure we had a full training support where it was a standalone impl not subject to future sudden breakage.

I appreciate the backward support, thus only commenting on the forward function file. :)

Let's discuss the rest of items internally, and I have a couple of concerns that we might not talk on the public repo.

Comment (lessw2020) [reply torchtitan/experiments/kernels/triton_group_gemm/tgroup_gemm_forward.py]: Sounds good - also M*G forward is up if you want to take a quick look (not perf optimized, just verified numerics for BF16):
https://github.com/pytorch/torchtitan/pull/960
